website/addons/box/model.py

```diff
@@ -0,0 +1,328 @@
+        return self.path.split('/')[1][-5] #self._metadata_cache['extra']['revisionId']
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Get or create a new file record. Return a tuple of the form (obj, created)
+        """
+        cleaned_path = clean_path(path)
+        try:
+            new = cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', cleaned_path)
+            )
+            created = False
+        except ModularOdmException:
+            # Create new
+            new = cls(node=node, path=cleaned_path)
+            new.save()
+            created = True
+        return new, created
+
+
+class BoxUserSettings(AddonUserSettingsBase):
+    """Stores user-specific box information, including the Oauth access
+    token.
+    """
+
+    box_id = fields.StringField(required=False)
+    access_token = fields.StringField(required=False)
+    refresh_token = fields.StringField(required=False)
+    box_info = fields.DictionaryField(required=False)
+    token_type = fields.StringField(required=False)
+    restricted_to = fields.DictionaryField(required=False)
+    last_refreshed = fields.DateTimeField(editable=True)
+
+
+    # TODO(sloria): The `user` param in unnecessary for AddonUserSettings
+    def to_json(self, user=None):
+        """Return a dictionary representation of the user settings.
+        The dictionary keys and values will be available as variables in
+        box_user_settings.mako.
+        """
+        output = super(BoxUserSettings, self).to_json(self.owner)
+        output['has_auth'] = self.has_auth
+        return output
+
+    def token_refreshed_callback(self, access_token, refresh_token):
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.last_refreshed = datetime.utcnow()
+        self.save()
+
+    @property
+    def has_auth(self):
+        return bool(self.access_token and refresh_creds_if_necessary(self))
+
+    def delete(self, save=True):
+        self.clear()
+        super(BoxUserSettings, self).delete(save)
+
+    def clear(self):
+        """Clear settings and deauthorize any associated nodes."""
+        self.box_id = None
+        self.access_token = None
+        for node_settings in self.boxnodesettings__authorized:
+            node_settings.deauthorize(Auth(self.owner))
+            node_settings.save()
+        return self
+
+    def get_credentialsv2(self):
+        return CredentialsV2(
+            self.access_token, 
+            self.refresh_token, 
+            settings.BOX_KEY,
+            settings.BOX_SECRET,
+            self.token_refreshed_callback,
+        )
+
+    def __repr__(self):
+        return u'<BoxUserSettings(user={self.owner.username!r})>'.format(self=self)
+
+
+class BoxNodeSettings(AddonNodeSettingsBase):
+
+    user_settings = fields.ForeignField(
+        'boxusersettings', backref='authorized'
+    )
+
+    folder = fields.StringField(default=None)
```

Comment:
Could the name of a folder and not the id change? If so we shouldn't be caching that property.